dhis-2/dhis-api/src/main/java/org/hisp/dhis/setting/LazySettings.java

```diff
@@ -0,0 +1,378 @@
+ *
+ * <p>Initially values are provided as raw {@link String} values.
+ *
+ * <p>When values are accessed as a specific type the conversion is applied to the {@link
+ * #rawValues} and remembered as {@link #typedValues}.
+ *
+ * @author Jan Bernitt
+ * @since 2.42
+ */
+@Slf4j
+@ToString
+@EqualsAndHashCode
+@JsonSerialize(using = LazySettings.SettingsSerializer.class)
+final class LazySettings implements SystemSettings, UserSettings {
+
+  private static final Set<String> CONFIDENTIAL_KEYS = new HashSet<>();
+  private static final Map<String, Serializable> DEFAULTS_SYSTEM_SETTINGS =
+      extractDefaults(SystemSettings.class);
+  private static final Map<String, Serializable> DEFAULTS_USER_SETTINGS =
+      extractDefaults(UserSettings.class);
+
+  private static final LazySettings EMPTY_USER_SETTINGS =
+      new LazySettings(UserSettings.class, new String[0], new String[0]);
+  private static final LazySettings EMPTY_SYSTEM_SETTINGS =
+      new LazySettings(SystemSettings.class, new String[0], new String[0]);
+
+  static Set<String> keysWithDefaults(Class<? extends Settings> type) {
+    if (type == SystemSettings.class) return Set.copyOf(DEFAULTS_SYSTEM_SETTINGS.keySet());
+    if (type == UserSettings.class) return Set.copyOf(DEFAULTS_USER_SETTINGS.keySet());
+    return Set.of();
+  }
+
+  static boolean isConfidential(@Nonnull String key) {
+    return CONFIDENTIAL_KEYS.contains(key);
+  }
+
+  @Nonnull
+  static LazySettings of(Class<? extends Settings> type, @Nonnull Map<String, String> settings) {
+    return of(type, settings, (k, v) -> v);
+  }
+
+  @Nonnull
+  static LazySettings of(
+      Class<? extends Settings> type,
+      @Nonnull Map<String, String> settings,
+      @Nonnull BinaryOperator<String> decoder) {
+    if (settings.isEmpty()) {
+      if (type == UserSettings.class) return EMPTY_USER_SETTINGS;
+      if (type == SystemSettings.class) return EMPTY_SYSTEM_SETTINGS;
+    }
+    // sorts alphabetically which is essential for the binary search lookup used
+    TreeMap<String, String> from = new TreeMap<>();
+    settings.forEach(
+        (k, v) -> {
+          if (v != null && !v.isEmpty()) from.put(k, v);
+        });
+    if (from.isEmpty()) {
+      if (type == UserSettings.class) return EMPTY_USER_SETTINGS;
+      if (type == SystemSettings.class) return EMPTY_SYSTEM_SETTINGS;
+    }
+    return of(type, from, decoder);
+  }
+
+  @Nonnull
+  private static LazySettings of(
+      Class<? extends Settings> type,
+      @Nonnull TreeMap<String, String> settings,
+      @Nonnull BinaryOperator<String> decoder) {
+    String[] keys = new String[settings.size()];
+    String[] values = new String[keys.length];
+    int i = 0;
+    for (Map.Entry<String, String> e : settings.entrySet()) {
+      String key = e.getKey();
+      String value = e.getValue();
+      keys[i] = key;
+      values[i++] = isConfidential(key) ? decoder.apply(key, value) : value;
```

Comment:
First time seeing an index used like this. This might not be clear to readers who don't know how pre/post increment works. Just an observation.

Reply:
Hehe, then it is good you haven't looked at some of the parser code I wrote :) But for clarity we can move the increment out to its own statement.